networking/ciao-cnci-agent/database.go

```diff
@@ -17,36 +17,196 @@
+	"time"
+
+	"github.com/01org/ciao/database"
 	"github.com/01org/ciao/payloads"
 	"github.com/golang/glog"
 )
 
-//Will implement a simple database to persist state across
-//restarts of the host/VM
+// cnciDatabase
+type cnciDatabase struct {
+	database.DbProvider //Database used to persist the CNCI state
+	SubnetMap
+	PublicIPMap
+}
+
+const (
+	tableSubnetMap   = "SubnetMap"
+	tablePublicIPMap = "PublicIPMap"
+)
+
+//dbCfg controls plugin data base attributes
+//these may be overridden by the caller if needed
+var dbCfg = struct {
+	Name    string
+	DataDir string
+	DbFile  string
+	Timeout time.Duration
+}{
+	Name:    "ciao cnci agent",
+	DataDir: "/var/lib/ciao/networking",
+	DbFile:  "docker_plugin.db",
+	Timeout: 1 * time.Second,
+}
+
+//SubnetMap maintains the list of active subnets across all compute nodes
+//that are handled by this CNCI
+type SubnetMap struct {
+	sync.Mutex
+	m map[string]*payloads.TenantAddedEvent //index: Agent IP + Subnet UUID
+}
+
+//NewTable creates a new map
+func (d *SubnetMap) NewTable() {
+	d.m = make(map[string]*payloads.TenantAddedEvent)
+}
+
+//Name provides the name of the map
+func (d *SubnetMap) Name() string {
+	return tableSubnetMap
+}
 
-func dbProcessCommand(client *agentClient, cmd *cmdWrapper) {
+//NewElement allocates and returns an subnet value
+func (d *SubnetMap) NewElement() interface{} {
+	return &payloads.TenantAddedEvent{}
+}
+
+//Add adds a value to the map with the specified key
+func (d *SubnetMap) Add(k string, v interface{}) error {
+	val, ok := v.(*payloads.TenantAddedEvent)
+	if !ok {
+		return fmt.Errorf("Invalid value type %t", v)
+	}
+	d.m[k] = val
+	return nil
+}
 
-	switch netCmd := cmd.cmd.(type) {
+//PublicIPMap maintains the list of active Public IP handled by this CNCI
+type PublicIPMap struct {
+	sync.Mutex
+	m map[string]*payloads.PublicIPCommand //index: PublicIP
+}
+
+//NewTable creates a new map
+func (d *PublicIPMap) NewTable() {
+	d.m = make(map[string]*payloads.PublicIPCommand)
+}
+
+//Name provides the name of the map
+func (d *PublicIPMap) Name() string {
+	return tablePublicIPMap
+}
+
+//NewElement allocates and returns an subnet value
+func (d *PublicIPMap) NewElement() interface{} {
+	return &payloads.PublicIPCommand{}
+}
+
+//Add adds a value to the map with the specified key
+func (d *PublicIPMap) Add(k string, v interface{}) error {
+	val, ok := v.(*payloads.PublicIPCommand)
+	if !ok {
+		return fmt.Errorf("Invalid value type %t", v)
+	}
+	d.m[k] = val
+	return nil
+}
+
+func dbInit() (*cnciDatabase, error) {
+	db := &cnciDatabase{}
+	db.DbProvider = database.NewBoltDBProvider()
+	db.SubnetMap.m = make(map[string]*payloads.TenantAddedEvent)
+	db.PublicIPMap.m = make(map[string]*payloads.PublicIPCommand)
+
+	if err := db.DbInit(dbCfg.DataDir, dbCfg.DbFile); err != nil {
+		return nil, err
+	}
+	if err := db.DbTableRebuild(&db.SubnetMap); err != nil {
+		return nil, err
+	}
+	if err := db.DbTableRebuild(&db.PublicIPMap); err != nil {
+		return nil, err
```

Comment:
return nil, errors.Wrap(err, "error rebuilding database tables for public IP map")